### ISSUE TITLE
NgForm.reset() always set a null value if the NgModel has a converter

### ISSUE THREAD

**Issue by anthony-foulfoin** [angular/angular.dart#1587]
Hi,

Given the following input in a form : 

```
<input type="text" ng-model="customer.birthay" date-formatter>
```

And the following decorator : 

```
@Decorator(selector: '[date-formatter]')
class DateFormatter implements AttachAware {

  NgModel model;

  DateFormatter(NgModel this.model) {
  }

  void attach() {
    model.converter = new DateConverter();
  }
}

class DateConverter implements NgModelConverter {

  DateFormat dateFormat = new DateFormat("d/M/y");

  @override
  parse(value) {
    return dateFormat.parse(value);
  }

  @override
  format(value) {
    return value == null ? null : dateFormat.format(value);
  }

}
```

The model is initialized with a value in the controller.
Then, I modify manually the input value, and I reset the form : 

```
form.reset();
```

I expect the model to be reset to its initialized value, but instead the model value is set to null.
If I remove the converter in the decorator, everything works well, the model is correctly reset to its initial value after resetting the form.

**Comment by matsko** [comment]
@tony75 I've been debugging the issue, but I'm unable to reproduce the issue in a test. Can you please let me know if I'm missing something: https://github.com/angular/angular.dart/commit/2b1d301f59f3dd6afa15b7502d07e1ac97c0b503

**Comment by tony75** [comment]
Maybe it's because your test method name starts with "itt" instead of "it"?

Le lun. 12 janv. 2015 00:12, Matias Niemelä <notifications@github.com> a
écrit :

**Comment by matsko** [comment]
`iit` is to have the test runner only focus on that one test instead of having to run everything each time I make a change.

**Comment by matsko** [comment]
Here is the updated commit: https://github.com/matsko/angular.dart/commit/2b1d301f59f3dd6afa15b7502d07e1ac97c0b503

**Comment by tony75** [comment]
Here is a very simple example that reproduces the problem : https://github.com/tony75/angulardartbug
If you run it and click twice on reset, a null value is set on the model.
If you remove the converter and click twice on reset, the initial value is set.

I don't just reset the form, I also flag it as submited : 

    form.reset();
    form.onSubmit(true);

It seems that the onSubmit(true) is the problem.

**Comment by matsko** [comment]
@tony75  On this line: https://github.com/angular/angular.dart/blob/master/lib/directive/ng_model.dart#L205 there's a try/catch statement which sets the model to be null in the event that the parser fails.

In the demo that you have it looks like that's what's causing the value to be set to null:

```
FormatException: Trying to read d from 2015-01-12 19:40:07.706 at position 0
```

The field should go blank right away after the first click. So that needs to be fixed. But the null value does behave as expected.

**Comment by tony75** [comment]
Thanks, I found what's causing the exception. I have updated the example to print the type of the value received in the converter parse method : 

```
  parse(value) {
    print(reflect(value).type.reflectedType.toString());
    return dateFormat.parse(value);
  }
```

When you modify the date in the input, the parse method receives a String as expected.

But when I click on reset, the parse methode receives a DateTime, which is the modelValue. I don't know if this is the expected behavior ?

**Comment by rkirov** [comment]
@matsko, I don't think your PR really solves the issue. After patching it and hitting `reset` twice, I still get the null value.

As I see it the issue is that `Converter.parse` happens to be called on the `_originalValue` after reset, but not at the initial load of the form.

I don't think we can guarantee that parse can always be called with String and it is ok to have the model for a `input[type=text]` to be a non-string object. The only requirement for `Converter.parse` is to be idempotent, that is `parse(parse(value))` to be `parse(value)`.

@tony75 for now the easiest solution would be to make your converter accept already parsed `DateTime` like this:
` return value is String ? dateFormat.parse(value) : value; `

I see it more of a documentation error, we should be clear on what is expected of converters.

**Comment by rkirov** [comment]
Maybe removing the `try-catch` around the `converter.parse` would be better as it would make the authors of converters realize the expected API faster.

**Comment by matsko** [comment]
@rkirov sorry for not being responsive on this the other day. Yes I think this PR needs more of an investigation to see if the bug really does exist in the way that we think.

**Comment by rkirov** [comment]
Yeah, no problem, this is trickier that originally expected for now I recommend an extra guard in the user code as I said above. We should take some extra time to fix it properly for v1.2 but I don't want to block 1.1 for it.
